app/consumer.py

```diff
@@ -1,33 +1,73 @@
 import logging
 from structlog import wrap_logger
 from app.async_consumer import AsyncConsumer
-from app.response_processor import ResponseProcessor
+from app.response_processor import ResponseProcessor, BadMessageError, RetryableError, DecryptError
 from app import settings
+from app.queue_publisher import QueuePublisher
+import sys
 
 logging.basicConfig(level=settings.LOGGING_LEVEL, format=settings.LOGGING_FORMAT)
 logger = wrap_logger(logging.getLogger(__name__))
 
 
+def get_delivery_count_from_properties(properties):
+    """
+    Returns the delivery count for a message from the rabbit queue. The
+    value is auto-set by rabbitmq.
+    """
+    delivery_count = 0
+    if properties.headers and 'x-delivery-count' in properties.headers:
+        delivery_count = properties.headers['x-delivery-count']
+    return delivery_count + 1
+
+
 class Consumer(AsyncConsumer):
+
+    def __init__(self, **kwargs):
+        super().__init__()
+        self.quarantine_publisher = QueuePublisher(logger, settings.RABBIT_URLS, settings.RABBIT_QUARANTINE_QUEUE)
+
     def on_message(self, unused_channel, basic_deliver, properties, body):
-        logger.info('Received message', queue=self.QUEUE, delivery_tag=basic_deliver.delivery_tag, app_id=properties.app_id)
+
+        delivery_count = get_delivery_count_from_properties(properties)
+
+        logger.info(
+            'Received message',
+            queue=self.QUEUE,
+            delivery_tag=basic_deliver.delivery_tag,
+            delivery_count=delivery_count,
+            app_id=properties.app_id
+        )
 
         processor = ResponseProcessor(logger)
-        options = ResponseProcessor.options()
 
         try:
-            message = body.decode("utf-8")
-            processed_ok = processor.process(message, **options)
+            processor.process(body.decode("utf-8"))
+            self.acknowledge_message(basic_deliver.delivery_tag, tx_id=processor.tx_id)
+
+        except DecryptError as e:
+            # Throw it into the quarantine queue to be dealt with
+            self.quarantine_publisher.publish_message(body)
+            self.reject_message(basic_deliver.delivery_tag, tx_id=processor.tx_id)
+            logger.error("Bad decrypt - quarantining", exception=e, tx_id=processor.tx_id, delivery_count=delivery_count)
 
-            if processed_ok:
-                self.acknowledge_message(basic_deliver.delivery_tag, tx_id=processor.tx_id)
+        except BadMessageError as e:
+            # If it's a bad message then we have to reject it
+            self.reject_message(basic_deliver.delivery_tag, tx_id=processor.tx_id)
+            logger.error("Bad message - rejected", exception=e, tx_id=processor.tx_id, delivery_count=delivery_count)
 
-        except Exception as e:
-            logger.error("ResponseProcessor failed", exception=e, tx_id=processor.tx_id)
+        except (RetryableError, Exception) as e:
```

Comment:
This means each item will get tried once initially, and when the queue circles back around to this message, it will be retired once again - rather than previous situations where there's been 5 retries before requeueing - if I understand correctly?